packages/neon-nep5/src/main.ts

```diff
@@ -141,3 +141,78 @@ export async function getToken(
     throw err;
   }
 }
+
+/**
+ * Retrieves the complete information about all (or a list of) tokens.
+ * @param url RPC Node url to query.
+ * @param scriptHashArray Array of NEP5 contract scriptHashes.
+ * @param address Optional address to query the balance for. If provided, the returned object will include the balance property.
+ */
+export async function getTokens(
+  url: string,
+  scriptHashArray: string[],
+  address?: string
+): Promise<TokenInfo[]> {
+  const sb = new sc.ScriptBuilder();
+  scriptHashArray.forEach(scriptHash => {
+    if (address) {
+      const addrScriptHash = u.reverseHex(
+        wallet.getScriptHashFromAddress(address)
+      );
+      sb.emitAppCall(scriptHash, "name")
+        .emitAppCall(scriptHash, "symbol")
+        .emitAppCall(scriptHash, "decimals")
+        .emitAppCall(scriptHash, "totalSupply")
+        .emitAppCall(scriptHash, "balanceOf", [addrScriptHash]);
+    } else {
+      sb.emitAppCall(scriptHash, "name")
+        .emitAppCall(scriptHash, "symbol")
+        .emitAppCall(scriptHash, "decimals")
+        .emitAppCall(scriptHash, "totalSupply");
+    }
+  });
+
+  try {
+    const res = await rpc.Query.invokeScript(sb.str).execute(url);
+
+    const result: TokenInfo[] = [];
+    const step = address ? 5 : 4;
+    for (let i = 0; i < res.result.stack.length; i += step) {
+      const name = rpc.StringParser(res.result.stack[i]);
+      const symbol = rpc.StringParser(res.result.stack[i + 1]);
+      const decimals = rpc.IntegerParser(res.result.stack[i + 2]);
+      const totalSupply = rpc
+        .Fixed8Parser(res.result.stack[i + 3])
+        .dividedBy(
+          Math.pow(10, decimals - rpc.IntegerParser(res.result.stack[i + 2]))
+        )
+        .toNumber();
+      const balance = address
+        ? rpc
+            .Fixed8Parser(res.result.stack[i + 4])
+            .dividedBy(
+              Math.pow(
+                10,
+                decimals - rpc.IntegerParser(res.result.stack[i + 2])
+              )
+            )
+        : undefined;
+
+      const obj = {
+        name,
+        symbol,
+        decimals,
+        totalSupply,
+        balance
+      };
+      
+      if(!obj.balance) delete obj.balance;
+
+      result.push(obj);
+    }
+    return result;
+  } catch (err) {
+    log.error(`getToken failed with : ${err.message}`);
+    throw err;

```

Comment:
I think it is weird that we only try catch the invocation and parsing part but not the script construction when you are going to log an generic error.

Either:

1. Try catch the entire thing to log the generic error.
2. Split the try catch into logical components, so one trycatch for scriptbuilder, one for invocation and one for parsing. 
3. Don't try catch at all since the `err.message` is likely to bubble all the way up and we are not adding any good contextual information to it here.

Reply:
I know this is likely because of the structure of `getTokenBalances` but I would like to take this chance to revisit error handling. Might have to invest some time into making it consistent throughout the codebase. I think you, as a dev working with it, is in a good spot to provide some feedback on what would improve your experience :)

Reply:
In my opinion, this is not the place where the "truly useful" are created.
If anything goes wrong on a deeper level (at the creation of the build script or something), this is the place where it should provide the necessary useful message(s) for the user to understand what went wrong. 

I think it's sufficient to have the underlying error (if the message is clear enough and if it's useful for the user) with the called function name contained in the error as done right here --> `log.error("getToken failed with : ${err.message}");`

Although this is just my sole opinion and better ways are 100% certainly possible. Maybe this is something to discuss on Slack/Discord instead.

As a last note, I've now move the try/catch up. If you'd like it to be reverted right now or applied to the other functions, I'm willing to do so!

Reply:
For debugging, I'm a big fan of grouping error messages by logical abstractions with verbosity to make it quick and easy to find a chunk of code that is problematic. It should be left up to project API implementations to make the errors user-friendly

Reply:
I just noticed that `getTokenBalances` has a try-catch in the loop and outside of it. Would you like me to nest it like that aswell?